### ISSUE TITLE
sql: scrutinize Kafka connection options

### ISSUE THREAD

**Issue by benesch** [MaterializeInc/materialize#14656]
We have a number of Kafka connection options that are suspect. In no particular order:

* `ACKS`: this doesn't seem like it ought to be configurable. Our correctness guarantees require "acks: all", and I don't think we should be allowing users to opt out of that correctness guarantee. Or maybe it doesn't matter at all when using the transactional producer... in which case that's a _different_, but still great, reason to get rid of this option.
* `ENABLE IDEMPOTENCE`: same as above. This shouldn't be configurable?
* `CLIENT ID`: now, this one seems reasonable! (From the Kafka docs: "The sole purpose of this is to be able to track the source of requests beyond just ip and port by allowing a logical application name to be included in Kafka logs and monitoring aggregates.") We might want to make this like `GROUP ID PREFIX`, where we only allow users to configure a prefix of the client ID, and always smush "Materialize" in there somehow. Not sure. Needs more thought.
* `PARTITION COUNT`, `REPLICATION FACTOR`, `RETENTION BYTES`, `RETENTION MS`: these control the configuration of the topic Materialize creates... if Materialize even creates a topic. If the topic already exists, Materialize is happy to use it as is, with whatever configuration it might have. I'm almost inclined to say that if you want to configure any properties about the Kafka topic that Materialize is writing to, you have to create that topic manually. If we want to keep the options, we need to document what the semantics are if the topic already exists in Kafka and the configuration of the topic in Kafka does not match the configuration of the topic in Materialize.
* `RETENTION MS`: we should use an interval type for this one.
* `TOPIC METADATA REFRESH INTERVAL MS`: this one should also be an interval type, and should be permanently unsafe. It exists only so that our tests can notice new partitions faster than every 5m.
* `STATISTICS INTERVAL MS`: this should evaporate; it controls how often Kafka statistics are emitted to the log file, but that's not accessible to users.

Filing under the surfaces project, but this will require a close collaboration with the storage team.

Re prioritization: as long as these options remain unsafe after #14600, there's no particular rush to solve this. We can pick them off one by one as customer requests come in.

cc @sploiselle @bkirwi

**Comment by bkirwi** [comment]
All makes sense! Quick thoughts:
- We probably want to set a client id, even if not explicitly configured... aside from the observability benefit, it's very useful for eg. configuring throttling. (If we do decide to let users configure it, I think it should be safe to let them put ~whatever they want?)
- Partition count and retention I'd be tempted to expose: extremely performance relevant, hard to pick a great default value for, shouldn't impact materialize's correctness.

**Comment by nmeagan11** [comment]
Below are some issues that have come up for users in support of bringing back some of these options:

- `PARTITION COUNT`, `REPLICATION FACTOR`, `RETENTION MS` ([reference](https://materializeinc.slack.com/archives/C01CFKM1QRF/p1664449089084219)). The current workaround One workaround is to manually create the topic in Kafka with the desired config, and just point Materialize to that. However, it would be nice to have these back to not have to do that.
- `CLIENT ID` ([reference](https://materializeinc.slack.com/archives/C0415G58EA3/p1664404884102729)). This is a nice-to-have in order to differentiate Materialize from other consumers.

**Comment by benesch** [comment]
I think we're probably behaving badly at the moment by not setting a good default for `client.id`. We should probably some up with something like `materialize-$ENVIRONMENT-$CONNID-$SRCID`.

**Comment by nmeagan11** [comment]
> `materialize-$ENVIRONMENT-$CONNID-$SRCID`

+1

**Comment by nmeagan11** [comment]
We recently merged a PR (https://github.com/MaterializeInc/materialize/pull/15609) that allows for a non-default partition count for an already-existing topic (which is the behavior we thought we had). Following this, I think it makes sense to bring back `replication_factor` and `partition_count`, but there's a product decision to be made on what behavior we want when the configuration is different from the option. 

For example, let's assume I have a pre-existing Kafka topic with a single partition, and I specify the `partition_count` to be 3. IMO the behavior we want is that we should ignore `partition_count` and take whatever is already specified in the pre-existing topic. We would need to send a `WARNING` to the user that their specified option will be ignored.

**Comment by benesch** [comment]
> For example, let's assume I have a pre-existing Kafka topic with a single partition, and I specify the `partition_count` to be 3. IMO the behavior we want is that we should ignore `partition_count` and take whatever is already specified in the pre-existing topic. We would need to send a `WARNING` to the user that their specified option will be ignored.

I think there are a few more things to shake out here:

  * Is there a way to specify that Materialize should create the default partition count besides omitting the `PARTITION COUNT` option entirely? In other places we use `0` to indicate the "default" value (e.g., `INTROSPECTION INTERVAL 0`); I believe Kafka might use `-1`? We could also introduce a sentinel like `DEFAULT`, which would be consist with how `SET` works in SQL.
  * When is the warning generated? Just on the initial `CREATE SOURCE`?
  * What happens if, some time after creating the source, the user changes the topic configuration in Kafka? Should Materialize periodically check for skew between the Kafka options provided to `CREATE SOURCE` and the actual configured topic options on the broker? Should it report that skew as some sort of warning in the new source errors collection?

Since the semantics here are subtle, I'd recommend that we start with the docs change rather than the implementation. Can we come up with a clear way of explaining the new behavior in the docs? I'll feel much better if we draft up that docs PR and feel good about it before embarking on the implementation.

**Comment by nmeagan11** [comment]
Taking a stab at some answers:

> Is there a way to specify that Materialize should create the default partition count

I think this might be a matter of being clear what our desired behavior is in the docs? E.g. if Materialize is creating the topic, we use X defaults.

> When is the warning generated?

I think how we display warnings in general is still being fleshed out (ccc @ggnall).

> What happens if, some time after creating the source, the user changes the topic configuration in Kafka?

Good point! Light thinking, but I suppose reporting it as an error works? Would we stop ingesting data if we detect a skew?

**Comment by benesch** [comment]
> > What happens if, some time after creating the source, the user changes the topic configuration in Kafka?
> 
> Good point! Light thinking, but I suppose reporting it as an error works? Would we stop ingesting data if we detect a skew?

Oops! I got this backwards! I meant "sink" not "source". But, no, we wouldn't by default stop exporting data if we detected a skew between the configuration in the original `CREATE SINK` and the current topic.

**Comment by benesch** [comment]
Some Slack discussion today around how `PARTITION COUNT` ought to work: https://materializeinc.slack.com/archives/C01CFKM1QRF/p1681924452485939

The proposal on the table is: make `PARTITION COUNT` a thing that is determined during purification. @bkirwi proposes that you *must* specify the partition count explicitly if the topic does not exist during purification. If the topic does exist, then I propose that purification adds the `PARTITION COUNT` explicitly to the statement, for clarity in `SHOW CREATE SOURCE`.

**Comment by benesch** [comment]
Closing this, as the prescribed scrutiny has occurred!

  * #23792 removes or fixes all remaining options listed in the issue description.
  * #23794 tracks the redesign and re-installation of the `PARTITION COUNT`, `REPLICATION FACTOR`, `RETENTION MS` and `RETENTION BYTES` options, which are removed by #23792.
